docs/quality_assurance.md

```diff
@@ -0,0 +1,157 @@
+- The correctness of the provider chain, i.e., the provider CCV module does not break the liveness or the safety of the provider chain. 
+- The correctness of the Interchain Security protocol, i.e., the protocol follows the [specification](https://github.com/cosmos/ibc/blob/master/spec/appics-028-cross-chain-validation/README.md).
+- The correctness of the consumer chain, i.e., both liveness or safety hold.
+
+> TODO decide what should be covered by the protocol audit and the priority
+
+## Software Engineering
+
+| Concern | Code Review | Automatic Tools | Unit Testing |
+| ------- | ----------- | --------------- | ------------ |
+| Unexpected panics | `TODO` | `??` | `??` |
+| Handling errors | `TODO` | `??` | `??` |
+| Accessing state (setters, getters, iterators) | `TODO` | `??` | `Low coverage` |
+| Serialization / deserialization | `TODO` | `??` | `??` |
+
+## Integration with IBC
+
+Interchain Security is an IBC application and, thus, it relies on IBC to establish a separate channel between the provider chain and every consumer chain. Interchain Security relies on the IBC v3.0 golang [implementation](https://github.com/cosmos/ibc-go/tree/v3.0.0).
+
+IBC packets:
+- ValidatorSetChangePacket
+- MaturedVSCPacket
+- SlashPacketData
+
+| Concern | Code Review | Unit Testing | Diff. testing | Testnet | Protocol audit |
+| ------- | ----------- | ------------ | ------------- | ------- | ----- |
+| Create IBC clients | `TODO` (ibc-go team) | `Done` | `??` | `TODO` | `NA` |
+| Getting consumer `UnbondingPeriod` from IBC client | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+| Create CCV channel (handshake) | `TODO` (ibc-go team) | `Done` | `??` | `TODO` | `TODO` |
+| Sending IBC packets <br /> - see `x/ccv/utils/utils.go:SendIBCPacket()` | `TODO` (ibc-go team) | `Done` | `??` | `TODO` | `NA` |
+| Handling acknowledgments | `TODO` (ibc-go team) | `Low coverage` | `??` | `TODO` | `NA` |
+| Handling timeouts | `TODO` (ibc-go team) | `Low coverage` | `??` | `TODO` | `NA` |
+| Handling IBC client expiration | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+| ICS-20 channel creation | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+| ICS-20 transfer | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+
+## Integration with Cosmos SDK
+
+A prerequisite of the code review is to open a PR with all the [SDK changes](https://github.com/cosmos/cosmos-sdk/tree/interchain-security-rebase) needed by Interchain Security.
+
+| Concern | Code Review | Unit Testing | Diff. testing | Testnet | Protocol audit |
+| ------- | ----------- | ------------ | ------------- | ------- | ----- |
+| Changes to staking module | `TODO` (sdk team) | `Low coverage` | `??` | `TODO` | `NA` |
+| Changes to slashing module | `TODO` (sdk team) | `??` | `??` | `TODO` | `NA` |
+| Changes to evidence module | `TODO` (sdk team) | `??` | `??` | `TODO` | `NA` |
```

Comment:
I didn't even know we had changes to slashing or evidence! Did this happen yet?

Reply:
Yes it did. The changes are mostly located in `x/slashing/keeper/infractions.go` and `x/evidence/keeper/infraction.go` and allow the CCV module to not necessarily have (for consumer chains) validator public keys and operator addresses stored in states.

Reply:
Great, thanks.